src/main.cpp

```diff
@@ -319,10 +319,10 @@ void UpdatePreferredDownload(CNode* node, CNodeState* state)
 }
 
 void InitializeNode(NodeId nodeid, const CNode *pnode) {
+    CAddress addr = pnode->addr;
+    std::string addrName = pnode->addrName;
     LOCK(cs_main);
-    CNodeState &state = mapNodeState.insert(std::make_pair(nodeid, CNodeState())).first->second;
-    state.name = pnode->addrName;
-    state.address = pnode->addr;
+    mapNodeState.emplace_hint(mapNodeState.end(), std::piecewise_construct, std::forward_as_tuple(nodeid), std::forward_as_tuple(addr, std::move(addrName)));
```

Comment:
IMO this change is an excessive optimization. The code is much less readable than the previous version and I doubt this code is a performance bottleneck. But I am willing to accept it to minimize divergence from upstream.